Dev btb rebase

jensen-yan opened the pull request:
rebase dev-btb to the master
use --bp-type=DecoupledBPUWithBTB
to use btb BPU

Lingrui98 committed:
btb-pred: add multi-way btb pred with ras and tage, other predictors to be added

Lingrui98 committed:
btb-pred: add ittage support

Lingrui98 committed:
btb-pred: should not match when global history differs

Lingrui98 committed:
btb-pred: fix btbEntriesWithDifferentStart

Lingrui98 committed:
btb-pred: use btb entries to decide whether pred hit when update

Lingrui98 committed:
pred: add btb-db

jensen-yan committed:
pred: support DecoupledBPUWithBTB

- Add support for DecoupledBPUWithBTB in xiangshan.py configuration
- Disable loop predictor, loop buffer, and jump ahead predictor for DecoupledBPUWithBTB
- Add enableSC parameter to BTBTAGE branch predictor

jensen-yan committed:
pred: Add Chinese README for XiangShan GEM5 branch predictor

- Add comprehensive README.cn.md for branch predictor module
- Provide detailed documentation for DecoupledBPUWithBTB architecture
- Include key data structures, prediction flow, and special features
- Add inline documentation for critical header files
- Create script for Kunming Lake 6-wide BTB configuration

jensen-yan committed:
pred: Refactor BTB predictors with unified numDelay parameter

- Remove hardcoded getDelay() methods in BTBTAGE, BTBITTAGE, DefaultBTB, and BTBRAS
- Add numDelay parameter to TimedBaseBTBPredictor base class
- Update BranchPredictor.py to use parent block size and set default numDelay values
- Simplify predictor configuration by using a common delay mechanism

jensen-yan committed:
btb-pred: fix compile bug using CC=clang

jensen-yan committed:
btb-pred: parameterize blockSize

not use fixed 32 byte or 0x20 byte,
now we can change it in xiangshan.py, use 64byte,
blockSize is equal to predictWidth using in FTB.

jensen-yan committed:
btb-pred: Add support for non-block-aligned BTB prediction

- Introduce `alignToBlockSize` parameter to control BTB prediction alignment
- Modify `stream_struct.hh` to handle fall-through address calculation based on alignment
- Update `btb.cc` to set index shift amount dynamically based on alignment
- Add configuration options in `xiangshan.py` for DecoupledBPUWithBTB
- Create script `kmh_v3_btb.sh` for BTB-based branch prediction configuration

eastonman commented:
Very difficult to review. May delay a few days if you want me to review the code.

jensen-yan committed:
util: Add compilation database generation support

- Usage: scons build/RISCV/compile_commands.json
- Ouput: in build/RISCV/compile_commands.json
- Enable SCons compilation database generation for all build variants
- Generate `compile_commands.json` in each variant directory
- Use absolute paths for better IDE integration

Change-Id: I1efb12ffd09c889235c5bea2860ca3d373701a50

jensen-yan committed:
btb-pred: Add μRAS unit test suite

- Implement comprehensive test suite for μRAS predictor using Google Test framework
- Add MockuRAS class to simulate core RAS functionality
- Create test cases covering:
  * Basic push and pop operations
  * Counter behavior
  * Stack full scenarios
  * Speculative update history
  * Prediction recovery mechanisms
- Add SConscript for test compilation
- Include detailed README.md explaining test suite structure and usage

This commit enhances the branch predictor testing infrastructure by providing
a robust test framework for the μRAS predictor.

jensen-yan committed:
btb-pred: Enhance BTB implementation with comprehensive documentation and comments

- Add detailed inline documentation for BTB implementation
- Improve code readability with extensive comments explaining:
  * BTB entry structure
  * Prediction and update mechanisms
  * MRU replacement policy
  * Address calculation methods
- Clarify complex logic in putPCHistory, lookup, and update methods
- Add comprehensive explanations for key functions and data structures
- Enhance code understanding for future maintainers

The changes provide a more transparent view of the Branch Target Buffer's internal workings without modifying core functionality.

jensen-yan committed:
btb-pred: Add comprehensive BTB unit test suite

- Implement MockDefaultBTB for testing Branch Target Buffer functionality
- Create BTB test suite using Google Test framework covering:
  * Basic initialization
  * Prediction with empty and updated BTB
  * Conditional branch prediction counter
  * Counter saturation behavior
  * MRU replacement policy
  * Indirect branch prediction
  * Multiple branch prediction
  * Misprediction recovery
- Add SConscript configuration for BTB test compilation
- Enhance testing infrastructure for branch predictor module

The changes provide a robust testing framework for the Branch Target Buffer implementation, improving code reliability and maintainability.

jensen-yan committed:
btb-pred: Refactor MockDefaultBTB with modular prediction logic

- Split putPCHistory method into smaller, focused functions:
  * processEntries: Sort and filter BTB entries
  * fillStagePredictions: Populate stage predictions
  * updatePredictionMeta: Manage prediction metadata
- Spict update method into more functions:
  * processOldEntries
  * get stats
  * collectEntries
  * updateBTBEntry
- Improve code modularity and readability
- Add detailed documentation for each new method
- Enhance separation of concerns in BTB prediction logic

The changes improve code structure without altering core functionality, making the MockDefaultBTB implementation more maintainable and easier to understand.

jensen-yan commented:
alignToBlockSize global variable setting have bugs!
I need to fix it!
I prefer to not use all global variables, I need to refactor stream_common.hh and blockSize.

jensen-yan committed:
util: CI: Add GitHub Actions workflow for Ideal BTB Performance Test

- Create new GitHub Actions workflow file `gem5-ideal-btb-perf.yml`
- Specify script path for Kunming Lake BTB performance test script

jensen-yan committed:
Merge branch 'xs-dev' into dev-btb-rebase

jensen-yan committed:
btb-pred: fix coredump, just for ci test

- Change `alignToBlockSize` default value from `true` to `false`
- Add TODO comment to remove global variable

Change-Id: I22e20c176724bcef0108bfa7eca927856706c8e9

jensen-yan committed:
btb-pred: Disable L0 BTB saturating counter update for performance investigation

- Temporarily disable BTB-specific conditional branch prediction mechanism
- This will cause specint score decrease from 18.6 to 18.0!

jensen-yan committed:
btb-pred: delete SC from TAGE

- Create btb_sc.hh and btb_sc.cc files for Statistical Corrector implementation
- Temporarily comment out SC implementation details
- Prepare infrastructure for future Statistical Corrector integration
- Remove commented-out SC code from btb_tage.cc and btb_tage.hh

This commit sets up the groundwork for adding a Statistical Corrector to the TAGE branch predictor, improving modularity and preparing for future enhancements.

jensen-yan committed:
btb-pred: Modularize DefaultBTB update method with fine-grained functions

- Refactor `update` method in DefaultBTB to improve code modularity
- Split complex update logic into smaller, focused functions:
  * `processOldEntries`: Filter and process old BTB entries
  * `checkPredictionHit`: Evaluate prediction hit status
  * `collectEntriesToUpdate`: Determine entries requiring updates
  * `updateBTBEntry`: Handle individual BTB entry updates
- Enhance code readability and maintainability
- Preserve existing BTB update semantics
- Add const-correctness to method parameters

This refactoring improves the DefaultBTB implementation's structure without changing its core functionality.

jensen-yan committed:
btb-pred: Enhance TAGE predictor with comprehensive code documentation and comment.

jensen-yan committed:
btb-pred: Align mockbtb.cc to btb.cc, it is dirty, but necessary for refactor btb.

jensen-yan committed:
btb-pred: Add comprehensive TAGE predictor unit test suite

- Implement comprehensive unit tests for BTBTAGE predictor using Google Test
- Cover key TAGE predictor functionalities:
  * Basic prediction mechanism
  * Branch history update and recovery
  * Folded history management
  * Main and alternative prediction selection
  * Speculative and actual branch outcome handling
- Add test cases for various scenarios:
  * Consecutive correct predictions
  * Alternating branch outcomes
  * Multiple branch sequences
  * Boundary conditions
- Enhance testing infrastructure for TAGE branch predictor
- Improve code reliability and predictor behavior understanding

The tests provide in-depth validation of the TAGE predictor's complex prediction logic and help ensure robust branch prediction performance.

jensen-yan committed:
btb-pred: Enhance TAGE predictor with comprehensive code documentation and debug logging

align btb_tage to test/btb_tage!

jensen-yan committed:
btb-pred: not use to_string in tage/ittage

to_string will case performance decrease
same as https://github.com/OpenXiangShan/GEM5/pull/158

jensen-yan committed:
btb-pred: fix unit test for tage

should update GHR first, then use newest GHR to recoverHist

jensen-yan committed:
btb-pred: refactor BTBTAGE

- Added helper methods for TAGE table lookup, single prediction generation, and entry preparation for updates.
- Improved the update process by modularizing the logic for predictor state updates and new entry allocation.
- Introduced structures for holding TAGE table lookup results and allocation results to streamline the code.
- Enhanced code documentation and debug logging for better maintainability and understanding of the prediction logic.

These changes aim to improve the overall structure and performance of the BTBTAGE predictor, aligning it with best practices in code organization and documentation.

boran-xu committed:
cpu-o3: ideal script now specify BPU type.

previously the script to simulate kmhv3 ideal does not specify BPU type, and the basic BPU configuration is used unintentionally.

Change-Id: I60816cfa28e939616c016ba45703b2a240760dcc

jensen-yan committed:
btb-pred: Rename MockDefaultBTB to DefaultBTB

- Align mockbtb.cc to btb.cc!
- Change Namespace mockBTB to test, same as tage.hh
- Changed all instances of MockDefaultBTB to DefaultBTB for consistency.
- Updated debug logging format specifiers from %d to %lu for size_t compatibility.

boran-xu committed:
btb-pred: typo in xiangshan.py, btb->ftb.

jensen-yan committed:
cpu-o3: Implement Half-Aligned BTB

- Added support for half-aligned mode in DefaultBTB, allowing 64-byte predictions while maintaining 32-byte alignment for index and tag calculations.
- Introduced new test cases to validate half-aligned mode functionality, including handling of unaligned addresses and multiple updates.
- Enhanced lookup and update mechanisms to accommodate the new half-aligned design, ensuring correct branch predictions across adjacent 32-byte blocks.
- Documented the design and testing strategy for the half-aligned BTB in halfAligned.md for better understanding and future reference.

Change-Id: I7fe7ee32f5b340a98040d9dbf317f4d7c10ab9bc

jensen-yan committed:
cpu-o3: btb-pred: Support Half-Aligned BTB really

- Copy from Mock BTB
- Updated DefaultBTB to support half-aligned mode, allowing for 64-byte predictions while maintaining 32-byte alignment for index and tag calculations.
- Introduced a new lookup method for single block entries to enhance the lookup process.
- Modified the update logic to accommodate half-aligned entries, ensuring correct handling of adjacent 32-byte blocks.
- Enhanced debug logging for better traceability during prediction operations.

These changes improve the flexibility and accuracy of the branch target buffer (BTB) in handling various alignment scenarios.

Change-Id: I44f07fd286276afae919b778628a34c80f83a10e

jensen-yan committed:
cpu-o3: btb-pred: fix bug: should update btb tag after found it

- since updateBTBEntry will find entry_to_write at first, then should update tag, should not update entry.tag first!
- Moved the implementation of getIndex and getTag methods inline for better readability and performance.
- Updated checkPredictionHit to remove unnessary code
- Modified updateBTBEntry to accept entryPC instead of btb_idx, enhancing clarity in the update process.
- Align mock btb to btb.

Change-Id: I1d1d960ceb0abdf304c5277d95b0372748441d70

jensen-yan committed:
cpu-o3: pred: Add design documentation and refactor BTB unit test

- Introduced a comprehensive design document for the DefaultBTB, detailing its architecture, key design principles, data structures, and algorithms.
- Added helper functions in the BTB test suite to facilitate the creation of branch information and setup of fetch streams for testing.
- Enhanced existing tests to utilize the new helper functions, improving readability and maintainability of the test code.
- Verified the functionality of the BTB in various scenarios, including half-aligned mode and multiple branch predictions.

These changes aim to improve the understanding and testing of the BTB implementation in gem5.

Change-Id: I87d69fc4bc2574421fdc66f944ec4a706a75cded

jensen-yan committed:
cpu-o3: pred: Enhance BTB and FetchStream initialization

- Updated BTBEntry constructor to initialize additional fields: alwaysTaken, ctr, and tag.
- Refactored FetchStream constructor to include new member variables: updateEndInstPC, squashPC, predTick, and history, improving clarity and initialization.
- Cleared various vectors in FetchStream to ensure proper memory management and prevent potential issues during execution.

These changes aim to improve the robustness and maintainability of the BTB and FetchStream components in the gem5 simulator.

Change-Id: I72879e1d18735f06f49b4751e96057d099eda79e

jensen-yan committed:
cpu-o3: pred: Refactor BTBTAGE unit test

- Introduced a new design document for the BTBTAGE
- Updated unit tests to utilize helper functions for better organization and clarity.

Change-Id: If8c36456993ed0e2519b8cdc234b17c2ed0a0d00

jensen-yan committed:
cpu-o3: pred: Refactor BTBTAGE lookup and prediction methods

- Replaced the TableLookupResult structure with a std::vector<TageEntry> for more efficient data handling during TAGE table lookups.
- Updated the lookupTageTable and generateSinglePrediction methods to reflect the new data structure, improving clarity and reducing complexity.

Change-Id: Iac9005c5dbbf0e1c02f53aa0b39454070951dd84

boran-xu committed:
btb-pred: added ahead pipeline feature on BTB, testing required.

jensen-yan committed:
cpu-o3: pred: add more TAGE unit test

- Modified BTBTAGE constructor to adjust table sizes, tag bits, PC shifts, and history lengths for improved prediction accuracy.
- Enhanced unit tests to validate the new design and ensure consistent prediction accuracy across various branch patterns.

Change-Id: I3ee33b12e24f863e024ece303574a4e165eeee55

boran-xu committed:
btb-pred: added abtb unit test, 1 passed, fixed bug

boran-xu committed:
btb-pred: change the name to abtb.

boran-xu committed:
btb-pred: pulled the calculation of tag and index out of updateBTBEntry()

Since we need to accommodate ahead pipelining, we need to have the possibility of calculating tag and index separately, using different PC,
thus we made this change that make the tag and index two arguments of the updateBTBEntry() function

boran-xu committed:
btb-pred: refactored abtb tests.

boran-xu committed:
btb-pred: decoupled global and local parameters(block size, half-aligned)

before there's only one global parameter that determined the alignment of sub-components, now we made it
such that all BTBs has a parameter entryHalfAligned that define their alignment.
Before there's a global variable blockSize at namespace 'btb-pred' that actually means predictWidth,
at the mean time each sub-components has a variable blockSize which is associated with its own internal design
To avoid confusion, we changed the name of the global variable to predictWidth

jensen-yan committed:
cpu-o3: pred: Enhance BTBTAGE with N-way Set Associative Design

- Added support for N-way set associative tables in the BTBTAGE predictor, allowing multiple entries per index to reduce conflicts and improve prediction accuracy.
- Introduced LRU (Least Recently Used) replacement policy to manage entries effectively, ensuring that frequently accessed entries are retained.
- Updated data structures to accommodate the new design, including modifications to the TageEntry and TageTableInfo structures.
- Refactored lookup and allocation methods to handle multiple ways, ensuring correct predictions and efficient entry management.
- Enhanced unit tests to validate the new design and ensure consistent performance across various scenarios.

Change-Id: I4b749b6e1b51d4bac96fc0e973de23864c74608c

jensen-yan committed:
cpu-o3: pred: add comment and unit tests for  Fetch Target Queue

- Added detailed documentation for FetchTargetQueue to facilitate understanding of the fetch target management process.
- Introduced unit tests for FetchTargetQueue to validate functionality, including enqueueing, fetching, and handling of edge cases.

Change-Id: Ib7d6e8c2034cc6967be5d6e5b8427bda0d425039

boran-xu committed:
btb-pred: integrated aBTB with BPU, bug need to be fixed.

after integrating aBTB with BPU, we encountered a bug where the Fetch Block predicted by aBTB
has branch that are outside the FB's limit. This is due to a type of Aliasing caused by two FB having different
index bits. I implemented a unit test to demonstrate this type of bug.

boran-xu reviewed (dismissed):
all looks fine, except that we should make MockBTB's alignToBlockSize parameterized. But I'll address that in my branch!

boran-xu committed:
cpu-o3: pred: fix aBTB bug by having tag cover all of PC.

when aBTB's tag only covers certain bits of a FB's Start_PC, it's prone to making inconsistent predictions (br_pc out of FB's limit). We fixed it by making tag start from the 2nd LSB and end at the MSB of Start_PC

Change-Id: I3380a724d4b55f78b3e545417706d5011a625889

boran-xu committed:
cpu-o3: pred: added option to use aBTB...

...simply add --abtb when running xiangshan.py

Change-Id: Ifc5b56c49723005dc1d11deeac6ed455606b6a6f

boran-xu committed:
cpu-o3: bpred: added a few options in run script

Change-Id: Id120fc71d6679e4485b0e89e48aa2fdbef806829

boran-xu committed:
cpu-o3: ahead pipeline debug flag.

Change-Id: I3bb0168cc8cebacb7b9fce5e63936a02597e23d7

boran-xu committed:
cpu-o3: pred: mockBTB: parameterization of half-alignment.

Change-Id: I672686538b7a695bb8d131a625b9a6d59e9d2516

jensen-yan committed:
cpu-o3: pred: Refactor BTB and TAGE to use TimedBaseBTBPredictor

- Updated BTBTAGE and DefaultBTB classes to inherit from TimedBaseBTBPredictor, enabling consistent delay management across predictors.
- This commit is prepared for adding DecoupledBPU unit test.

boran-xu committed:
cpu-o3: pred: btb override statistic counters.

Change-Id: I77cccacb73b53b4d452bbe547dd2ce39de991bc2

jensen-yan committed:
cpu-o3: build: Add UNIT_TEST macro in SConstruct

- Introduced a new command-line option '--unit-test' in SConstruct to enable unit test builds.
- This is used when compile unit test, replace debug/DecoupleBP.hh with test_dprintf.hh

Change-Id: I2d305811a0a4d9dfdf5de80feca793edb3c90f68

jensen-yan committed:
cpu-o3: pred: fix abtb unit test after rebase

Change-Id: I961f626ee88e6b185030f2d6fea9537dfac46d70

boran-xu committed:
cpu-o3: pred: added statistics that tracks override reason.

Change-Id: I2507f96f6d73f0e5883ab089cc603a63572c1466

jensen-yan committed:
cpu-o3: pred: Update debug includes for unit testing

- Modified jump_ahead_predictor.hh, loop_buffer.hh, and loop_predictor.hh to include "test_dprintf.hh" when UNIT_TEST is defined, enhancing support for unit testing.
- Adjusted DPRINTF statements to use %ld for size_t types to ensure correct formatting.

jensen-yan committed:
cpu-o3: pred: Implement DecoupledBPUWithBTB unit test framework

- Introduced the DecoupledBPUWithBTB mock class.
- Splict HistoryManager from DecoupledBPU.hh
- Created unit tests for DecoupledBPUWithBTB, covering basic prediction now.
- Updated SConscript to include new test files and ensure proper unit testing setup.

jensen-yan committed:
cpu-o3: pred: Update DecoupledBPUWithBTB unit tests

- Added a new constructor for BranchInfo to facilitate testing.
- Introduced mock PCState and StaticInst classes for enhanced unit testing capabilities.
- Updated unit tests to cover the new pipeline stages and prediction processes.

Change-Id: Ie0b5e19dbedd3f28132371e5ef2da120bc49cb99

jensen-yan committed:
cpu-o3: pred: Enhance Jump Ahead Predictor with unit tests

- Added comprehensive comments and documentation for the Jump Ahead Predictor class and its associated structures.
- Implemented unit tests for the Jump Ahead Predictor to validate its functionality and performance.
- Updated SConscript to include the new test file for the Jump Ahead Predictor.

Change-Id: I845891cf9735ad5025e14fa8537c5ae10e639300

boran-xu committed:
cpu-o3: pred: ubtb and abtb use the same range as mbtb

Change-Id: I2d9bcd8baf0b6b1752b6a8e4e84e4c0c3bf6b96f

jensen-yan committed:
cpu-o3:pred: Enhance DecoupledBPUWithBTB with detailed comments

- Added debug statements to track the prediction process, including when the fetch stream queue is full, final predictions are generated, and override bubbles are consumed.
- Refactored the `generateFinalPredAndCreateBubbles` method to improve clarity and maintainability, including detailed logging for each step of the prediction process.
- Updated the `tryEnqFetchStream` and `tryEnqFetchTarget` methods to include additional validation and logging for better debugging.

boran-xu committed:
cpu-o3: pred: add perf stats: commitOverrideBubble.

Change-Id: I22a1ce504006e59344db31cef6dece13c68393ed

boran-xu committed:
cpu-o3: fix UpdateInstEndPC in fetch stream

the fall through address should be + predictWidth then 32B aligned. Not + predictWidth then 64B aligned.

Change-Id: Ic1463501740ce8bcf0a3d5ff234baafdaba9bb75

jensen-yan committed:
cpu-o3: pred: Refactor BTBBPU to remove loop-related/jap

- Removed unused loop predictor and loop buffer components from DecoupledBPUWithBTB, simplifying the class structure.
- Cleaned up related statistics and methods that were dependent on loop prediction functionality.
- Updated fetch target queue and stream structures to eliminate loop iteration tracking.
- New Loop Predictor and jap waiting to be added later!

Change-Id: Ifc700f675b284ed59bceb30c8c61ead8faab4280

jensen-yan committed:
cpu-o3: pred: Refactor BPU for improved validation and structure

- Refactored the `tryEnqFetchStream` and `tryEnqFetchTarget` methods to include validation functions `validateFSQEnqueue` and `validateFTQEnqueue`, enhancing code clarity and maintainability.
- Updated the `setTakenEntryWithStream` method signature for consistency.
- Added detailed comments to new validation functions and existing methods to improve documentation and understanding of the prediction process.
- Enhanced the `createFetchStreamEntry` and `updateHistoryForPrediction` methods to streamline the prediction entry creation and history update processes.

Change-Id: I55a650514c4881ad1600bbfe8d9a9384b1ab2a82

jensen-yan committed:
cpu-o3: pred: Refactor DecoupledBPUWithBTB tick method

- Enhanced the `tick` method by breaking it down into helper functions: `processEnqueueAndBubbles` and `requestNewPrediction`, improving code readability and maintainability.
- Added detailed comments to clarify the purpose of each step in the prediction cycle.

Change-Id: I0cb290f3a06ca4eec5a77e5b8789e8a9ace0faf0

jensen-yan committed:
cpu-o3: pred: Refactor squash handling in BPU

- Introduced a new `handleSquash` method to encapsulate common logic for handling different types of squash events, improving code organization and readability.
- Updated `controlSquash`, `nonControlSquash`, and `trapSquash` methods to utilize the new `handleSquash` method, reducing code duplication.
- Enhanced comments and documentation to clarify the purpose and functionality of the squash handling process.
- Delete unused numbr

Change-Id: I113c3b42ff60ffb00430220b40f5fa4b84faab0b

jensen-yan committed:
cpu-o3: pred: Refactor history recovery and update methods in BPU

- Introduced a new method `recoverHistoryForSquash` to encapsulate the logic for recovering branch history during squash events, improving code organization and readability.
- Enhanced comments for history_manager

Change-Id: I0689fe56f6da5d7ab2f9140afd08bc55445f5b64

jensen-yan committed:
cpu-o3: pred: Refactor BPU update methods

- Introduced helper functions `updatePredictorComponents` and `updateStatistics` to encapsulate logic for updating predictor components and statistics, improving code organization and readability.
- Enhanced comments for clarity on the purpose of each function and the overall update process.

Change-Id: I2c1e3d4a71851db1741979cd9bdf7c492d7d9152

jensen-yan committed:
cpu-o3: pred: Refactor BPU decoupledPredict method

- Updated the `decoupledPredict` method to improve handling of fetch target predictions, including better management of instruction counters and clearer logic for taken and not-taken branches.
- Introduced a new helper function `processFetchTargetCompletion`

Change-Id: If62bc310fe06a8a9f603faca94f19e47f73df7ca

boran-xu committed:
cpu-o3: pred: align btb counter initialization with ftb.

boran-xu committed:
cpu-o3: pred: fixed processOldEntries logic in uBTB.

before, uBTB's entriesToUpdate get indiscriminately pushed back stream.updateNewBTBEntry. it causes duplication that result in certain counters getting updated twice, creating a huge increase of ubtb.condPredWrong and thus override bubble. this is fixed by removing duplication.

Change-Id: I72b1a20ac2cb6cd95cd3acdab461e139a15b2845

jensen-yan committed:
cpu-o3: pred: fix bug: Update maxShamt from 8 to 16 in foldedHist

- Since in 64Byte FetchBundle, may have more than 8 branches, especially in gobmk, maxShamt = 8 will cause check history failed.
- Updated `FoldedHist` initialization to use 16 bits instead of 8 for `tagFoldedHist`, `altTagFoldedHist`, and `indexFoldedHist`.
- 16 may needs to be parameterred, but we will use PHR later, GHR + foldedHist will be removed.

Change-Id: I2a775d8679c0d395198e36f742fdca6a0e8f61d0

jensen-yan committed:
cpu-o3: pred: Refactor BPU parameters and remove unused files

- Removed the `alignToBlockSize` parameter from `DecoupledBPUWithBTB` and replaced it with `predictWidth` for better clarity and functionality.
- Updated various methods to utilize the new `predictWidth` parameter for fall-through and target address calculations.
- Deleted unused `stream_common.cc` and `stream_common.hh` files to clean up the codebase and reduce redundancy.

Change-Id: I0c852a18b0b99247bde1e926ad9d91993bcab5f9

jensen-yan committed:
cpu-o3: pred: Enhance BTB statistics tracking

- Added new statistics for tracking existing entries updated, entries replaced, and entries replaced with valid entries in `DefaultBTB`.
- Updated various methods to accommodate the new statistics and ensure accurate tracking of fetch stream and target queue states.
- Use maxInstsNum instead of hard-coded 16

jensen-yan committed:
Merge branch 'xs-dev' into dev-btb-rebase

boran-xu committed:
cpu-o3: pred: add override source statistics for FTB

Change-Id: I5875beef80ed0ed4b22b64ba224ea6abf9ddb5a7

boran-xu committed:
cpu-o3: pred: move override reason statistics to commit path.

Change-Id: Id5763ade2286363d730d091d6837cc018b566edd

jensen-yan committed:
cpu-o3: pred: Refactor BPU statistics tracking

- Introduced a new method `dumpStats` to centralize the output of various branch prediction statistics, including mispredictions and BTB entries.
- Reorganized existing code to improve clarity and maintainability, including the addition of helper functions for file creation and sorting.
- Updated the `commitBranch` method to utilize the new `processMisprediction` function for better handling of misprediction statistics.
- Enhanced the `BranchStats` structure to provide detailed tracking of misprediction types and rates.
- Improved documentation throughout the code to clarify the purpose and functionality of new and existing methods.

This refactor aims to streamline the tracking and reporting of branch prediction performance metrics, facilitating better analysis and debugging.

Change-Id: I12cdc8a78b7af8646e8af431943b9871763bb0a8

boran-xu committed:
cpu-o3: pred: fixed bug: S1pred return address is not assigned.

When uBTB predicts a return branch, the getTarget() function of FullBTBPrediction uses the returnTarget member variable as the target. Which isn't assigned a value. this results in a considerable increase of S2 flush, because the target doesn't match with that of RAS. This commit fixed it.

Change-Id: I02ba91577c4f403ddc41faacb8eaf0627bb20d9b

jensen-yan committed:
cpu-o3: pred: add more bpu trace, enhance Fetch stage documentation

- Updated the GitHub Actions workflow to trigger only on the `xs-dev` branch for push and pull requests.
- Added detailed documentation for the Fetch stage in `fetch.md`, covering its interfaces, data structures, and core functions.
- Improved code comments in `fetch.hh` to clarify the purpose of various member variables and structures.
- Enhanced BTB prediction statistics tracking in `btb_tage.cc` and `decoupled_bpred.cc`, including new trace records for better analysis.
- Useage: add --enable-bp-db basic predfsq predftq btb_0 tage, then sqlite3 bp.db --init GEM5/util/arch_db/bptrace.sql

eastonman commented on configs/example/fs.py:
```diff
@@ -318,15 +318,24 @@ def build_test_system(np):
+                            print("loop predictor not supported for this branch predictor")
+                        if args.enable_jump_ahead_predictor:
+                            print("jump ahead predictor not supported for this branch predictor")
+                    # assert(False)
```
This part is strange to me. Is this necessary?

eastonman commented on configs/example/xiangshan.py:
```diff
@@ -116,15 +116,26 @@ def build_test_system(np, args):
-        if args.bp_type is None or args.bp_type == 'DecoupledBPUWithFTB':
+        if args.bp_type is None or args.bp_type == 'DecoupledBPUWithFTB' or args.bp_type == 'DecoupledBPUWithBTB':
             enable_bp_db = len(args.enable_bp_db) > 1
             if enable_bp_db:
                 bp_db_switches = args.enable_bp_db[1] + ['basic']
                 print("BP db switches:", bp_db_switches)
             else:
                 bp_db_switches = []
-
-            test_sys.cpu[i].branchPred = DecoupledBPUWithFTB(
+            # for DecoupledBPUWithBTB, loop predictor and jump ahead predictor are not supported
+            #if args.bp_type == 'DecoupledBPUWithBTB':
+            if True:
+                if args.enable_loop_predictor or args.enable_loop_buffer:
+                    print("loop predictor and loop buffer not supported for DecoupledBPUWithBTB")
+                    args.enable_loop_predictor = False
+                    args.enable_loop_buffer = False
+                if args.enable_jump_ahead_predictor:
+                    print("jump ahead predictor not supported for DecoupledBPUWithBTB")
+                    args.enable_jump_ahead_predictor = False
+
+            BPClass = DecoupledBPUWithBTB() if args.bp_type == 'DecoupledBPUWithBTB' else DecoupledBPUWithFTB()
+            test_sys.cpu[i].branchPred = BPClass(
```
Maybe just add one more branch args.bp_type == 'decoupledBPUWithBTB' here.
`if True` here is quite dirty.

boran-xu replied:
fixed

eastonman commented on configs/example/xiangshan.py:
```diff
@@ -348,18 +359,31 @@ def setKmhV3IdealParams(args, system):
+                # TODO: BTB TAGE do not bave base table, do not support SC
+                cpu.branchPred.tage.tableSizes = [4096] * 14  # BTB TAGE may need larger table
+                if args.huge_ubtb:
+                    cpu.branchPred.abtb.numDelay = 9
+                    cpu.branchPred.ubtb.numEntries = 1024
+                elif args.pure_ubtb:
+                    cpu.branchPred.abtb.numDelay = 9
+                elif args.pure_abtb:
+                    cpu.branchPred.ubtb.numDelay = 9
```
What is `9`?

boran-xu replied:
added explanatory comments

eastonman commented:
Why prefetcher/cmc is modified?
I suggest split those small fixes into seperate PR, including `.gitignore` etc.

boran-xu committed:
cpu-o3: code style adjustment at BPU portion of python script.

Change-Id: I7fa5988484021eb1f2e5edda004c277da35fddfa

jensen-yan commented:
We have split some commits into different PRs, python scripts are also fixed.

eastonman reviewed (approved):
LGTM
